### ISSUE TITLE
[Arch] Backintime fails to build

### ISSUE THREAD

**Issue by afettouhi** [bit-team/backintime#1333]
Hi, I have an up to date Arch install and I am trying install the git version of backintime via the AUR using yay. But the install fails during unittesting for reason. Here is the output when it fails

OK
python3 -m unittest  -b test/test_backintime.py
E..
======================================================================
ERROR: test_diagnostics_arg (test.test_backintime.TestBackInTime)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/test/test_backintime.py", line 181, in test_diagnostics_arg
    diagnostics = json.loads(output)
  File "/usr/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

----------------------------------------------------------------------
Ran 3 tests in 3.040s

**Comment by aryoda** [comment]
Thanks for reporting this, I am already aware of this bug in another issue (no real duplicate):
See https://github.com/bit-team/backintime/issues/1233#issuecomment-1279248296

The current work-around to install from AUR is described here:
https://github.com/bit-team/backintime/issues/921#issuecomment-1276888138
It would be great if you report back here if it works or something else is missing...

**Comment by aryoda** [comment]
Fix submitted via PR #1335 (also updates the AUR installation section in README incl. a link for a work-around to this issue)

**Comment by afettouhi** [comment]
Still fails to build for me on Arch. Though the error is different now

`python3 -m unittest  -b test/test_backintime.py
Traceback (most recent call last):
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/backintime.py", line 1190, in <module>
    startApp()
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/backintime.py", line 507, in startApp
    args = argParse(None)
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/backintime.py", line 568, in argParse
    args, unknownArgs = mainParser.parse_known_args(args)
  File "/usr/lib/python3.10/argparse.py", line 1859, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/usr/lib/python3.10/argparse.py", line 2068, in _parse_known_args
    start_index = consume_optional(start_index)
  File "/usr/lib/python3.10/argparse.py", line 2008, in consume_optional
    take_action(action, args, option_string)
  File "/usr/lib/python3.10/argparse.py", line 1936, in take_action
    action(self, namespace, argument_values, option_string)
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/backintime.py", line 742, in __call__
    diagnostics = collect_diagnostics()
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/diagnostics.py", line 80, in collect_diagnostics
    result['host-setup']['locale'] = ', '.join(locale.getlocale())
TypeError: sequence item 0: expected str instance, NoneType found
E.s
======================================================================
ERROR: test_diagnostics_arg (test.test_backintime.TestBackInTime)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/test/test_backintime.py", line 191, in test_diagnostics_arg
    output = subprocess.check_output(["./backintime", "--diagnostics"])
  File "/usr/lib/python3.10/subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['./backintime', '--diagnostics']' returned non-zero exit status 1.

----------------------------------------------------------------------
Ran 3 tests in 2.966s

FAILED (errors=1, skipped=1)
make: *** [Makefile:570: unittest] Error 1
make: Leaving directory '/home/af/.cache/yay/backintime-git/src/backintime/common'`

**Comment by emtiu** [comment]
Thanks for reporting, I'm re-opening the ticket and assigning to @aryoda :)

**Comment by aryoda** [comment]
@afettouhi I did not test it with `yay` but differently. Could you please describe the exact steps to reproduce this (mainly the`yay` call with all arguments)?

**Comment by afettouhi** [comment]
I am just trying to build this way with yay

yay -S backintime-cli-git backintime-git

**Comment by aryoda** [comment]
The error tracebacks are exactly reproducible!

**Comment by afettouhi** [comment]
So, is this an issue with yay then?

**Comment by aryoda** [comment]
> So, is this an issue with yay then?

No it isn't at all. I could reproduce the issue by cloning BiT from the master branch on fresh Arch installation.

I am preparing a fix for this and will explain the details then here

**Comment by buhtzz** [comment]
Do I see it correct that `locale.getlocale()` return `None` on an Arch system?

**Comment by afettouhi** [comment]
[af@andre ~]$ locale
LANG=da_DK.UTF-8
LC_CTYPE="da_DK.UTF-8"
LC_NUMERIC="da_DK.UTF-8"
LC_TIME="da_DK.UTF-8"
LC_COLLATE="da_DK.UTF-8"
LC_MONETARY="da_DK.UTF-8"
LC_MESSAGES="da_DK.UTF-8"
LC_PAPER="da_DK.UTF-8"
LC_NAME="da_DK.UTF-8"
LC_ADDRESS="da_DK.UTF-8"
LC_TELEPHONE="da_DK.UTF-8"
LC_MEASUREMENT="da_DK.UTF-8"
LC_IDENTIFICATION="da_DK.UTF-8"
LC_ALL=

**Comment by aryoda** [comment]
Sorry to revise my first statement that I can reproduce the exact traceback. I have different ones in nearly the same locations.

I am still working on a fix for those but the originally reported traceback I cannot reproduce:

> File "/home/af/.cache/yay/backintime-git/src/backintime/common/diagnostics.py", line 80, in collect_diagnostics
    result['host-setup']['locale'] = ', '.join(locale.getlocale())
TypeError: sequence item 0: expected str instance, NoneType found

@buhtzz Are you working on that (just to be sure we don't do the same work)? I think the diagnostics code requires some more try-catches for such edge cases...

The currently known reasons to me for the other failing unit tests are:

1. Missing dependencies on fresh install in PKGBUILD:
     `sudo pacman -Sy python-packaging` (required by BiT) + `sudo pacman -Sy python-pyfakefs` (required for a unit test)
     -> I would like to contact all maintainers of ARCH BiT packages since this problem looks universal
     
2. Error in implementation of our new feature `--diagnostics`:
     RTE when using **new rsync version** that supports -VV to write version information as JSON
     because BiT tries a string trim on a dict (created from JSON) in that case instead of trimming a version string
     -> fix already prepared but not yet published
     
3. Unit test bug that I myself have introduced by changing the log output (write all log to `stderr`, incl. INFO and WARNING)
     https://github.com/bit-team/backintime/pull/1337
     One unit test now fails in case of a first-time installation because the BiT serviceHelper D-Bus daemon
     is not yet installed and BiT throws a different sequence of warnings than the unit test ist ignoring
     (the regexp ignores any output at the beginning and end of the log output, but not within the expected output)
     -> I will prepare a fix the next days to make the expected output regexp more tolerant

The reason why the BiT installation from AUR succeeded in my tests a few days ago are clear now to me:
- The old `rsync` version was still installed where the version number is still a string (not a JSON via `-VV`).
  I should have updated all my Arch packages first...
- New unit test bugs were introduced after my successful test via the "log to stderr" PR which was merged later

**Comment by buhtzz** [comment]
> @buhtzz Are you working on that (just to be sure we don't do the same work)?

No I don't.

> I think the diagnostics code requires some more try-catches for such edge cases...

Maybe. But even the `getlocal()` should return something. If the return value is `None` then something is very wrong with the underlying system.
We can't catch every possible error.

> 1. Missing dependencies on fresh install in PKGBUILD:
> ...
>    -> I would like to contact all maintainers of ARCH BiT packages since this problem looks universal

Why there are multiple maintainers on one distro?

We need to clearly separate the responsibilities and resources between upstream and distros.

**Comment by buhtzz** [comment]
> [af@andre ~]$ locale LANG=da_DK.UTF-8

Strange. There is a locale but while PKGBUILD there is no locale.

**Comment by aryoda** [comment]
> Strange. There is a locale but while PKGBUILD there is no locale.

PKGBUILD may possibly run as `sudo` or even `root` so @afettouhi could you please also check the locale as `sudo` (or whatever privileges `yay` is started...

**Comment by aryoda** [comment]
And what does `python3 -c "import locale; print(locale.getlocale())"` say? I'd expect `None` to be consistent with the unit test failure...

**Comment by afettouhi** [comment]
[af@andre ~]$ python3 -c "import locale; print(locale.getlocale())"
('da_DK', 'UTF-8')

**Comment by aryoda** [comment]
And

```
python3 -c "import locale; print(locale.getlocale(locale.LC_MESSAGES))"
python3 -c "import locale; print(locale.getlocale(locale.LC_CTYPE))"
```

**Comment by buhtzz** [comment]
Thanks to that thread in the arch forum
https://bbs.archlinux.org/viewtopic.php?pid=2063725#p2063725

We knew that the PKGBUILD "script" does set `LC_ALL=C` explicit before calling `make`.
Now I understand. And it is OK for me. It isn't a "bug" in the PKGBUILD script.

**Comment by afettouhi** [comment]
Glad I could help.

**Comment by aryoda** [comment]
I have fixed all problems (I hope) and contacted the AUR package maintainer to indicate open issues at the packaging side.

@afettouhi  Once the PR is merged into BiT you could try to install it again and it should work now (or I have work again ;-)

**Comment by afettouhi** [comment]
Looking forward to the update.

**Comment by willemw12** [comment]
All tests in https://aur.archlinux.org/pkgbase/backintime-git are passing now.
